example/src/screens/JoinRoom.tsx

```diff
@@ -0,0 +1,235 @@
+import { TextColors } from '@colors';
+import { BackgroundAnimation } from '@components/BackgroundAnimation';
+import { Modal } from '@components/Modal';
+import { TextInput } from '@components/TextInput';
+import { Typo } from '@components/Typo';
+import { StandardButton } from '@components/buttons/StandardButton';
+import { RootStack } from '@model/NavigationTypes';
+import { useHeaderHeight } from '@react-navigation/elements';
+import type { NativeStackScreenProps } from '@react-navigation/native-stack';
+import { useCardAnimation } from '@react-navigation/stack';
+import {
+  checkIfUrl,
+  extractRoomNameFromUrl,
+  shouldEnableRoomButton,
+} from '@utils';
+import React, { useEffect, useState, useCallback, useRef } from 'react';
+import {
+  View,
+  StyleSheet,
+  KeyboardAvoidingView,
+  Platform,
+  PermissionsAndroid,
+  ScrollView,
+} from 'react-native';
+import { useVideoroomState } from 'src/VideoroomContext';
+
+type Props = NativeStackScreenProps<RootStack, 'JoinRoom'>;
+type GoBackAction = Readonly<{
+  type: string;
+  payload?: object | undefined;
+  source?: string | undefined;
+  target?: string | undefined;
+}>;
+
+export const JoinRoom = ({ navigation, route }: Props) => {
+  const height = useHeaderHeight();
+  const [isModalVisible, setIsModalVisible] = useState(false);
+  const modalAction = useRef<GoBackAction>();
+  const { roomName, setRoomName, username, setUsername } = useVideoroomState();
+  const [isRoomNameInputEditable, setIsRoomNameInputEditable] = useState(true);
+  const { next, current } = useCardAnimation();
+
+  useEffect(() => {
+    if (route.params?.roomName) {
+      setRoomName(route.params?.roomName);
+      setIsRoomNameInputEditable(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    const handleBeforeRemoveEvent = (e) => {
+      if (!roomName && !username) {
+        // If we don't have unsaved changes, then we don't need to do anything
+        return;
+      }
+      e.preventDefault();
+      modalAction.current = e.data.action;
+      setIsModalVisible(true);
+    };
+
+    navigation.addListener('beforeRemove', handleBeforeRemoveEvent);
+
+    return () =>
+      navigation.removeListener('beforeRemove', handleBeforeRemoveEvent);
+  }, [navigation, roomName, username]);
+
+  useEffect(() => {
+    navigation.setOptions({
+      headerStyle: {
+        //@ts-ignore
+        opacity: next
+          ? 0
+          : current.progress.interpolate({
+              inputRange: [0, 0.99, 1],
+              outputRange: [0, 0, 1],
+            }),
+      },
+    });
+  }, []);
+
+  const openPreview = () => {
+    if (checkIfUrl(roomName)) {
+      setRoomName(decodeURI(extractRoomNameFromUrl(roomName)));
+    }
+    navigation.push('Preview', { title: 'Join meeting' });
+  };
+
+  const requestPermissionsAndOpenPreview = useCallback(async () => {
```

Comment:
This function should be extracted somewhere else, even to a separate file, you're repeating it.

Reply:
Done.